value letrec: fix a bug in analysis of inner mutually-recursive bindings

gasche opened the pull request:
On Friday I found a bug in my own proof of correctness of the recursive-value-declarations analysis of #1942, and @yallop swiftly turned it into a example that crashes 4.08+dev. The present PR (explains and) fixes the issue in the same way that I fixed the type system and the proof.

The issue is in the handling of inner `let rec` bindings with mutual recursion going on (single-definitions `let rec` are fine). Read the comments and testsuite entries for the details. There are two potential fixes, a fixpoint (precise/complete, slightly more complex to implement), and an overapproximation (rejects safe programs, shorter to implement). The paper, the new proof and this PR use the fixpoint approach.

The new unsafe program in the testsuite is rejected in 4.07.1, which still uses @yallop's pre-#1942 implementation, which is more conservative -- it is equivalent to the "overapproximation" mentioned above. On the other hand, the new safe program in the testsuite was rejected in 4.07.1, because the overapproximation is too conservative there -- maybe it's okay.

We should have a fix for this issue in 4.08.

gasche committed:
value letrec: fix a bug in analysis of inner mutually-recursive bindings

Jeremy Yallop and myself found this bug while trying to prove the
soudness of the analysis.

gasche commented:
@yallop: would you be able to review this PR in the next few weeks?

yallop commented:
@gasche: I think it's quite unlikely.

gasche commented:
Thanks; I'll try to find another victim.

Armael commented on typing/rec_check.ml:
```diff
@@ -1030,22 +1043,82 @@ and extension_constructor : Typedtree.extension_constructor -> term_judg =
+           (Gi, (xj : mij)^j |- ei : m[mi])^i   (xi : mi -| D)^i
+           G'i = Gi + mij[G'j]
+           --------------------------------------------------------
+           Sum(G'i) + (D - (pj)^j) |- let rec (xi=ei)^i : m -| D
```
Naming of variables/patterns seems inconsistent: `xi` vs `pj`

Armael commented on typing/rec_check.ml:
```diff
@@ -1030,22 +1043,82 @@ and extension_constructor : Typedtree.extension_constructor -> term_judg =
+              List.map mode bindings in
+            (* Gi *)
+            let outer_binding_env = remove_patlist all_bound_pats rhs_env in
+            (* (Gi, (mij)^j *)
```
closing parenthesis missing

Armael reviewed (commented):
A couple typos.

gasche commented:
@Armael and @damiendoligez agreed to review the PR. I just fixed a couple of typos spotted by @Armael, and implemented a different, hopefully less-confusing naming schemes for multi-index families of modes.

damiendoligez commented on typing/rec_check.ml:
```diff
@@ -1030,22 +1043,88 @@ and extension_constructor : Typedtree.extension_constructor -> term_judg =
+          (* [binding_env] takes a binding (x_i = e_i)
+             and computes (Gi, (mdef_ij)^j). *)
+          let binding_env {vb_pat = p_i; vb_expr = e_i; _} =
+            (* m[mi] *)
```
This `mi` should be `mbody_i`.

gasche replied:
Thanks! I made some minor changes to the naming that fixed this. This part now reads as follows:

```ocaml
          (* [binding_env] takes a binding (x_i = e_i)
             and computes (Gi, (mdef_ij)^j). *)
          let binding_env {vb_pat = x_i; vb_expr = e_i; _} =
            let mbody_i = pattern x_i bound_env in
            (* Gi, (x_j:mdef_ij)^j  *)
            let rhs_env_i = expression e_i (Mode.compose mode mbody_i) in
            (* (mdef_ij)^j (for a fixed i) *)
            let mutual_modes =
              let mdef_ij {vb_pat = x_j; _} = pattern x_j rhs_env_i in
              List.map mdef_ij bindings in
            (* Gi *)
            let env_i = remove_patlist all_bound_pats rhs_env_i in
            (* (Gi, (mdef_ij)^j) *)
            (env_i, mutual_modes) in
```

damiendoligez reviewed (approved):
After getting a white-board explanation of the bug and the fix from @gasche, I had a good look at the code and I believe this patch is correct.
P.S. the new naming scheme is indeed much better.

gasche commented:
Thanks to the reviews of @Armael and @damiendoligez, I will merge and cherry-pick in 4.08 once the CI turns green again.

gasche commented:
Cherry-picked as 8d7d7580b3b052cbffd19ae1d0b0232e8fa0a789.
